apps/src/lab2/levelEditors/validations/EditValidations.tsx

```diff
@@ -0,0 +1,164 @@
+
+/**
+ * This component is used to edit validations for a level. Currently only used
+ * by Lab2 labs.
+ */
+
+const AppConditions: {[key in AppName]?: ConditionType[]} = {
+  music: Object.values(MusicConditions),
+};
+
+interface EditValidationsProps {
+  initialValidations: Validation[];
+  levelName: string;
+  appName: AppName;
+}
+
+/**
+ * Top-level validations editor component.
+ */
+const EditValidations: React.FunctionComponent<EditValidationsProps> = ({
+  initialValidations,
+  levelName,
+  appName,
+}) => {
+  const [validations, setValidations] = useState<Validation[]>(
+    sanitizeValidations(initialValidations, levelName)
+  );
+
+  const conditionTypes = AppConditions[appName];
+
+  if (conditionTypes === undefined || conditionTypes.length === 0) {
+    return (
+      <div>
+        {`No validation conditions available for app type: ${appName}. Contact the engineering team for further details.`}
+      </div>
+    );
+  }
+
+  const onValidationChange = (validation: Validation) => {
+    setValidations(
+      sanitizeValidations(
+        validations.map(v => {
+          if (v.key === validation.key) {
+            return validation;
+          }
+          return v;
+        }),
+        levelName
+      )
+    );
+  };
+
+  const deleteValidation = (key: string) => {
+    setValidations(
+      sanitizeValidations(
+        validations.filter(validation => {
+          return validation.key !== key;
+        }),
+        levelName
+      )
+    );
+  };
+
+  const moveValidation = (key: string, direction: 'up' | 'down') => {
+    const index = validations.findIndex(validation => {
+      return validation.key === key;
+    });
+
+    if (index === -1) {
+      return;
+    }
+
+    const newIndex = direction === 'up' ? index - 1 : index + 1;
+    if (newIndex < 0 || newIndex >= validations.length) {
+      return;
+    }
+    const newValidations = [...validations];
+    const temp = newValidations[index];
+    newValidations[index] = newValidations[newIndex];
+    newValidations[newIndex] = temp;
+    setValidations(newValidations);
+  };
+
+  const addValidation = () => {
+    const newValidation: Validation = {
+      key: levelName + '_' + createUuid(),
+      message: '',
+      next: false,
+      conditions: [],
+    };
+    setValidations([...validations, newValidation]);
+  };
+
+  return (
+    <div>
+      <input
+        type="hidden"
+        id="level_validations"
+        name="level[validations]"
+        value={JSON.stringify(validations)}
+      />
+      <Typography
+        semanticTag="p"
+        visualAppearance="body-three"
+        className={moduleStyles.title}
+      >
+        Edit validations for this level. Currently only supported by Lab2 labs.
+        <br />
+        NOTE: Validations are checked in the order they are listed. The first
+        validation set that passes is the one that is displayed, so be sure to
+        order your validation sets from most conditions to least.
```

Comment:
This could possibly be "most stringent to least" or similar, since it might be the case that we check for one validation earlier even though it technically doesn't have more conditions than those below it?